src/main/java/org/elasticsearch/index/translog/Translog.java

```diff
@@ -1500,14 +1530,225 @@ public Durabilty getDurabilty() {
          */
         REQUEST;
 
-        public static Durabilty getFromSettings(ESLogger logger, Settings settings, Durabilty defaultValue) {
-            final String value = settings.get(INDEX_TRANSLOG_DURABILITY, defaultValue.name());
-            try {
-                return valueOf(value.toUpperCase(Locale.ROOT));
-            } catch (IllegalArgumentException ex) {
-                logger.warn("Can't apply {} illegal value: {} using {} instead, use one of: {}", INDEX_TRANSLOG_DURABILITY, value, defaultValue, Arrays.toString(values()));
-                return defaultValue;
+    }
+
+    private static void verifyChecksum(BufferedChecksumStreamInput in) throws IOException {
+        // This absolutely must come first, or else reading the checksum becomes part of the checksum
+        long expectedChecksum = in.getChecksum();
+        long readChecksum = in.readInt() & 0xFFFF_FFFFL;
+        if (readChecksum != expectedChecksum) {
+            throw new TranslogCorruptedException("translog stream is corrupted, expected: 0x" +
+                    Long.toHexString(expectedChecksum) + ", got: 0x" + Long.toHexString(readChecksum));
+        }
+    }
+
+    public static Snapshot snapshotFromStream(StreamInput input, final int numOps) {
+        final BufferedChecksumStreamInput checksumStreamInput = new BufferedChecksumStreamInput(input);
+        return new Snapshot() {
+            int read = 0;
+            @Override
+            public int estimatedTotalOperations() {
+                return numOps;
+            }
+
+            @Override
+            public Operation next() throws IOException {
+                if (read < numOps) {
+                    read++;
+                    return readOperation(checksumStreamInput);
+                }
+                return null;
+            }
+
+            @Override
+            public void close() {
+                // doNothing
+            }
+        };
+    }
+
+    static Translog.Operation readOperation(BufferedChecksumStreamInput in) throws IOException {
+        int opSize = in.readInt();
+        Translog.Operation operation;
+        try {
+            in.resetDigest(); // size is not part of the checksum?
+            if (in.markSupported()) { // if we can we validate the checksum first
+                in.mark(opSize);
+                if (opSize < 4) { // 4byte for the checksum
```

Comment:
Can we move this if + exception to the top of the method, just after reading opSize?